lucille-core/src/main/java/com/kmwllc/lucille/indexer/OpenSearchIndexer.java

```diff
@@ -96,9 +104,107 @@ protected void sendToIndex(List<Document> documents) throws Exception {
       return;
     }
 
-    BulkRequest.Builder br = new BulkRequest.Builder();
+    Map<String, Document> documentsToUpload = new LinkedHashMap<>();
+    Set<String> idsToDelete = new LinkedHashSet<>();
+    Map<String, List<String>> termsToDeleteByQuery = new LinkedHashMap<>();
 
+    // populate which collection each document belongs to
+    // if document is marked for deletion ONLY, then only add to idsToDelete
+    // if document is marked for deletion AND contains deleteByFieldField and deleteByFieldValue, only add to termsToDeleteByQuery
+    // else then add to upload
     for (Document doc : documents) {
+      String id = doc.getId();
+      if (isMarkedForDeletion(doc)) {
+        documentsToUpload.remove(id);
+        if (isMarkedForDeletionByField(doc)) {
+          String field = doc.getString(deleteByFieldField);
+          if (!termsToDeleteByQuery.containsKey(field)) {
+            termsToDeleteByQuery.put(field, new ArrayList<>());
+          }
+          termsToDeleteByQuery.get(field).add(doc.getString(deleteByFieldValue));
+        } else {
+          idsToDelete.add(id);
```

Comment:
could rearrange the conditional so the simplest clauses is the first one

Reply:
resolved in [af56102](https://github.com/kmwtechnology/lucille/pull/191/commits/af56102e1864b5297de99dfa81a184026483d666)